### PR TITLE
Refactorize Scope

### PR DESCRIPTION
**Pull request by wangkuiyi**
Fixes https://github.com/PaddlePaddle/Paddle/issues/3113
Fixes https://github.com/PaddlePaddle/Paddle/issues/3114
Fixes https://github.com/PaddlePaddle/Paddle/issues/3115
Fixes https://github.com/PaddlePaddle/Paddle/issues/3117

### CODE REVIEW

**Comment by wangkuiyi** [comment]
`recurrent_network_op.{h,cc}` is using the `shared_ptr` syntax of the current `Scope` implementation. I need to change this part to use `Scope::NewScope` before this PR could pass CI.

File: paddle/framework/scope.cc [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    return iter->second;
+  }
+  Variable* v = new Variable();
+  vars_[name] = v;
+  v->name_ = &(vars_.find(name)->first);
+  return v;
+}
+
+Variable* Scope::NewVar() {
+  return NewVar(string::Sprintf("%p.%d", this, vars_.size()));
```
Comment (jacquesqiao) [inline paddle/framework/scope.cc]: It seems use `vars_.size()` as the unique key is not so good? why not `atmoic` or `timestamp`?

File: doc/design/scope.md [hunk 2]
```diff
@@ -107,11 +107,11 @@ class Scope {
 ```
 ## Only scope can create a variable
 
-To ensure `only scope can create a variable`, we should mark `Variable`'s constructor as a private member function, and Scope is a friend class of Variable. And then only `CreateVariable` can construct `Variable`.
+To ensure `only scope can create a variable`, we should mark `Variable`'s constructor as a private member function, and Scope is a friend class of Variable. And then only `NewVar` can construct `Variable`.
```
Comment (Superjom) [inline doc/design/scope.md]: use a macro like `TF_DISALLOW_COPY_AND_ASSIGN` better? https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/macros.h#L84

Comment (Superjom) [inline doc/design/scope.md]: no only `Variable`, `Operator`, `Tensor`, `Scope` are all `DISALLOW_COPY_AND_ASSIGN`, this macro will make default and copy constructor `= delete`.

File: paddle/framework/scope.h [hunk 3]
```diff
@@ -35,73 +35,33 @@ class Scope;
...
 
  private:
-  std::unordered_map<Variable*, std::string> var_to_name_;
-  std::unordered_map<std::string, std::unique_ptr<Variable>> name_to_var_;
-  std::shared_ptr<Scope> parent_{nullptr};
+  // Call Scope::NewScope for a sub-scope.
+  explicit Scope(Scope* parent) : parent_(parent) {}
+
+  std::map<std::string, Variable*> vars_;
+  std::list<Scope*> kids_;
```
Comment (jacquesqiao) [inline paddle/framework/scope.h]: why should we store kid scope and how can we use them? 
In this PR, it seems that we will have a global default scope and all other scopes will be stored and `deleted` by this Scope if itself is deleted.

File: paddle/framework/scope.cc (hunk 1 above)
Comment (wangkuiyi) [inline paddle/framework/scope.cc]: I think scope address + the offset of the variable in this scope would be enough to uniquely identify the variable, isn't it?

File: paddle/framework/scope.h (hunk 3 above)
Comment (wangkuiyi) [inline paddle/framework/scope.h]: You are right. The purpose of adding`kids_` is to manage the lifecycle of Scopes.

**Pull request merged**